[ldinghan] iP

ldinghan opened the pull request:
# Ding, Duke's assistant
> "He's a little confused, but he got the spirit." - Will Smith ([source](https://knowyourmeme.com/memes/he-a-little-confused-but-he-got-the-spirit))

Ding cleans up after your mess. It's
* Quirky
* Text-based
* Easy to learn
* ~FAST~ _SUPER_ FAST to use

All you need to do is,

1. Download it from [here](https://github.com/ldinghan/ip/releases/download/A-Jar/duke.jar).
1. Double-click it.
1. Add your tasks.
1. Let it manage your tasks for you 🤯 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ldinghan committed:
Level-0

ldinghan committed:
Level-1

ldinghan committed:
Level-2

ldinghan committed:
Level-3

ldinghan committed:
Level-4

ldinghan committed:
A-TextUiTesting

ldinghan committed:
A-Exceptions

ldinghan committed:
Level-5

ldinghan committed:
Level-6

ldinghan committed:
A-Enums

ldinghan committed:
update javadocs

ldinghan committed:
Add Level-7

ldinghan committed:
Merge branch 'branch-Level-7'
Add Save functionality

ldinghan committed:
Add Level-8

ldinghan committed:
Merge branch 'branch-Level-8'

Switch datetime inputs to LocalDate and LocalTime types.

ldinghan committed:
A-MoreOOP

ldinghan committed:
Merge branch 'branch-A-MoreOOP'

Lacks OOP design.

Refactor code to enhance OOP design.

This improves the OOP design.

ldinghan committed:
Organise files into packages

ldinghan committed:
Merge branch 'branch-A-Packages'

Does not have package structure.

Adds package structure.

The files are more organised now.

ldinghan committed:
Merge branch 'add-gradle-support'

Add gradle support.

ldinghan committed:
Add Gradle support

ldinghan committed:
Add Junit Tests

ldinghan committed:
Modify Gradle Build file

ldinghan committed:
Add JavaDoc comments

ldinghan committed:
Merge branch 'branch-A-JavaDoc'
Add JavaDocs.

ldinghan committed:
Follow Coding Standard

ldinghan committed:
Merge branch 'branch-A-CodingStandard'
Did not follow Coding Standard.

Follow Coding Standard.

Code will follow Coding Standard.

ldinghan committed:
Add Find functionality

ldinghan committed:
Merge branch 'branch-Level-9'
Add Find functionality.

ldinghan committed:
Add 'find' in help command list

EricXiong420 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+package duke;
+
+import duke.commands.Command;
+import duke.exceptions.*;
```
Should this be split into multiple imports (instead of wildcard)? Wildcard imports aren't allowed in the course, and will be flagged out by CheckStyle

EricXiong420 commented on src/main/java/duke/exceptions/InvalidDateTimeException.java:
```diff
@@ -0,0 +1,16 @@
+    public InvalidDateTimeException(String message) {
+        super(message);
+    }
+}
```
Missing new line after the last line. This is present in a couple of other files too. Will be flagged up by CheckStyle.

EricXiong420 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,164 @@
+
+import duke.commands.Command;
+import duke.exceptions.*;
+import duke.tasks.*;
```
Similar to previous: shouldn't be using wildcard imports. Consider importing only as needed!

pzl111 replied:
Agreed, but I think there might be some careless autocorrecting by IntelliJ? On my version of IntelliJ, auto importing all the classes within a package will automatically collapse the import statement to use *?

EricXiong420 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,141 @@
+            }
+            File taskFile = new File(filePath);
+            if (taskFile.createNewFile()) {
+                //new file created
```
Consider adding a space after the "//", may appear neater.

EricXiong420 reviewed (commented):
Overall, I found your code easy to read for the most part except a few places where the concatenation of strings was a little confusing. Consider using something like String.format() to make those sections more readable. I noted a few minor coding standard violations too. The splitting into classes was well done!

pzl111 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,141 @@
+                    }
+                    taskList.addTask(taskT);
+                    break;
+                case 'D':
+                    description = description.split(" \\(by: ")[0];
+                    String by = str.split(" \\(by: ")[1];
+                    LocalDate date = LocalDate.parse(by.split(" ")[0]);
+                    LocalTime time = LocalTime.parse(by.split(" ")[1].replaceAll(".$", ""));
+                    Deadline taskD = new Deadline(description, date, time);
+                    if (isMarked) {
+                        taskD.markAsDone();
+                    }
+                    taskList.addTask(taskD);
+                    break;
+                case 'E':
+                    description = description.split(" \\(from: ")[0];
+                    String from = String.join("", str.split("\\(from: ")[1]).split(" to: ")[0];
+                    String to = str.split(" to: ")[1];
+                    LocalDate fromDate = LocalDate.parse(from.split(" ")[0]);
+                    LocalTime fromTime = LocalTime.parse(from.split(" ")[1]);
+                    LocalDate toDate = LocalDate.parse(to.split(" ")[0]);
+                    LocalTime toTime = LocalTime.parse(to.split(" ")[1].replaceAll(".$", ""));
+                    Event taskE = new Event(description, fromDate, fromTime, toDate, toTime);
```
Maybe you could consider extracting out the functionalities within each case into its separate functions? So the switch block will not be so long as it could be hard to trace the different cases when they are so many lines apart?

pzl111 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+            try {
+                Command command = Parser.getCommand(str);
+                switch (command) {
+                case LIST:
+                    Ui.printListReply(tasks);
+                    break;
+                case MARK:
+                case UNMARK:
+                    Ui.markTaskReply(Parser.taskToMark(str, tasks), tasks, command);
+                    break;
+                case TODO:
+                case DEADLINE:
+                case EVENT:
+                    tasks.addTask(Parser.parseStringToTask(str));
+                    Ui.addTaskReply(str, tasks);
+                    break;
+                case DELETE:
+                    Ui.deleteTaskReply(Parser.taskToDelete(str, tasks), tasks);
+                    break;
+                case FIND:
+                    Ui.findTaskReply(str, Parser.findKeyword(str, tasks));
+                    break;
+                case BYE:
+                    isRunning = false;
+                    Ui.exitReply();
+                    return;
```
Perhaps you could consider handling the different cases separately rather than using waterfall? Might be hard to trace the flow of the program, even though waterfall will reduce code duplication

pzl111 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,164 @@
+                throw new InvalidDescriptionException("Invalid description.");
+            }
+        case DEADLINE:
+            if (str.split(" ").length > 3) {
+                String fullTaskDescription = str.split(" ", 2)[1];
+                String description = fullTaskDescription.split(" /by ")[0];
+                String by = fullTaskDescription.split(" /by ")[1];
+                String[] datetime = by.split(" ");
+                try {
+                    LocalDate date = LocalDate.parse(datetime[0]);
+                    LocalTime time = LocalTime.parse(datetime[1]);
+                    Deadline deadline = new Deadline(description, date, time);
+                    return deadline;
+                } catch (DateTimeParseException e) {
+                    throw new InvalidDateTimeException("Invalid Datetime.");
+                }
+            } else {
```
Perhaps you could look into making the Happy Path prominent? swap out the if statement to test for the opposite, so that the unusual case is nested instead

pzl111 commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,45 @@
+            Task deadline = Parser.parseStringToTask("deadline Assignment 0 /by 2023-09-04 23:59");
+            assertEquals(new Deadline("Assignment 0", LocalDate.parse("2023-09-04"),
+                            LocalTime.parse("23:59")).toString(), deadline.toString());
+        } catch (InvalidDescriptionException e) {
+        } catch (InvalidDateTimeException e) {
+        }
```
I don't think there should be any empty catch blocks used? if so, at least provide some explanation for others to understand?

pzl111 reviewed (commented):
Overall good work, just some places could have used the Happy Path approach

ldinghan committed:
Add CheckStyle

ldinghan committed:
Fix exception in main file

ldinghan committed:
Fix style

ldinghan committed:
Add GUI

ldinghan committed:
Merge branch 'branch-Level-10'

Lacks GUI.

This merge adds GUI.

This will improve user experience.

ldinghan committed:
Merge conflicts

ldinghan committed:
Fix Level-10 and CheckStyle

ldinghan committed:
Improve Code Quality

Code does not follow Code Quality guidelines.

Improving Code Quality will enhance readability for future developments.

Code is being changed to ensure that Code Quality is improved.

Doing so will make the subsequent development easier.

ldinghan committed:
Merge pull request #1 from ldinghan/branch-A-CodeQuality

Improve Code Quality

ldinghan committed:
Use Assertions

Code does not use assertions.

Using assertions will act as a secondary barrier to prevent runtime
errors.

Code is changed to use assertions to ensure that possible runtime errors
are reduced.

Doing so will make subsequent development easier.

ldinghan committed:
Merge branch 'master' into branch-A-Assertions

ldinghan committed:
Merge pull request #3 from ldinghan/branch-A-Assertions

Use Assertions

ldinghan committed:
Fix Parser

ldinghan committed:
Add Snooze function

ldinghan committed:
Merge branch 'branch-B-Snooze'

Lacks Snooze functionality.

This merge adds Snooze functionality.

This will allow users to reschedule tasks.

ldinghan committed:
Add User Guide

ldinghan committed:
Make GUI unresizeable

ldinghan committed:
Update User Guide
